### PR TITLE
Teams feature

### PR DESCRIPTION
**Pull request by isht3**
```
* evalai teams
* evalai teams create
* evalai challenge CHALLENGE participate TEAM
```

### CODE REVIEW

File: evalai/challenges.py [hunk 1]
```diff
@@ -7,6 +7,13 @@
                                     display_past_challenge_list,
                                     display_participated_or_hosted_challenges,)
 
+from evalai.utils.teams import challenge_participate
+
+
+class Challenge(object):
```
Comment (RishabhJain2018) [inline evalai/challenges.py]: Please add docstring here.

File: evalai/challenges.py [hunk 2]
```diff
@@ -26,6 +33,16 @@ def challenges(ctx, participant, host):
         display_all_challenge_list()
 
 
+@click.group()
+@click.pass_context
+@click.argument('CHALLENGE', type=int)
+def challenge(ctx, challenge):
+    """
+    Challenge related operations.
```
Comment (RishabhJain2018) [inline evalai/challenges.py]: This is very general, please add specific docstring related to function.

File: evalai/challenges.py [hunk 3]
```diff
@@ -54,3 +71,15 @@ def future():
...
+
+@challenge.command()
+@click.pass_obj
+@click.argument('TEAM', type=int)
+def participate(ctx, team):
+    """
+    Participate in a challenge.
+
+    Invoked by running `evalai challenge CHALLENGE participate TEAM`
+    """
```
Comment (RishabhJain2018) [inline evalai/challenges.py]: **One issue:** 
In the current docstring, the `--help` command isn't able to display the complete docstring.
**Fix:**
Increase the word limit of `--help` command. Please check this if we can alter this. if not, then change the docstring to this format.
```
"""
Participate in a challenge
"""
"""
Invoked by running `evalai challenge CHALLENGE participate TEAM`
"""
```
The first comment here will be used by the `--help` command whereas second one will help any other developer or user.

File: evalai/utils/teams.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import sys
+
+from click import echo, style
+
+from evalai.utils.auth import get_request_header
+from evalai.utils.common import validate_token
+from evalai.utils.urls import URLS
+
+
+EVALAI_ERROR_CODES = [400, 401, 406]
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Please define it in `config.py`

File: evalai/utils/teams.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+API_HOST_URL = os.environ.get("EVALAI_API_URL", 'http://localhost:8000')
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Its already defined in `config.py`. No need to define it again & again.

File: evalai/utils/teams.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def pretty_print_teams(teams):
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Please change it to `pretty_print_team_data`

File: evalai/utils/teams.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    Pretty prints the teams
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: `Function to print the team data`

File: evalai/utils/teams.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    for team in teams:
+        br = style("----------------------------------------"
+                   "--------------------------", bold=True)
+
+        team_name = "\n{}".format(style(team["team_name"],
+                                        bold=True, fg="green"))
+        team_id = "ID: {}\n\n".format(style(str(team["id"]),
+                                            bold=True, fg="blue"))
+        title = "{} {}".format(team_name, team_id)
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: please change it to `team_name`

File: evalai/utils/teams.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        created_by = "Created by : {}\n\n".format(style(team["created_by"], fg="blue"))
+        members = "{}\n".format(style("Members", bold=True))
+        for member in team["members"]:
+            members = "{}* {}\n".format(members, member["member_name"])
+        team = "{}{}{}\n{}".format(title, created_by, members, br)
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: please change `title` to `team_name`

File: evalai/utils/teams.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        echo(team)
+
+
+def display_participant_teams():
+    """
+    Function to fetch all the participant teams of a user.
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: ` Function to display all the participant teams of a user`

File: evalai/utils/teams.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    headers = get_request_header()
+
+    url = "{}{}".format(API_HOST_URL, URLS.participant_team_lists.value)
+
+    try:
+        response = requests.get(url, headers=headers)
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        if (response.status_code in EVALAI_ERROR_CODES):
+            echo(style("Error: " + response.json()["error"], fg="red", bold=True))
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: use `.format` here.

File: evalai/utils/teams.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    for team in teams:
+        br = style("----------------------------------------"
+                   "--------------------------", bold=True)
+
+        team_name = "\n{}".format(style(team["team_name"],
+                                        bold=True, fg="green"))
+        team_id = "ID: {}\n\n".format(style(str(team["id"]),
+                                            bold=True, fg="blue"))
+        title = "{} {}".format(team_name, team_id)
+        created_by = "Created by : {}\n\n".format(style(team["created_by"], fg="blue"))
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Can we please change the color from blue. It doesn't appears to be very clear and good.

File: evalai/utils/teams.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    except requests.exceptions.HTTPError as err:
+        if (response.status_code in EVALAI_ERROR_CODES):
+            echo(style("Error: " + response.json()["error"], fg="red", bold=True))
+        else:
+            echo(err)
+        sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        echo(err)
+        sys.exit(1)
+    response_json = response.json()
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Change it to `response`

File: evalai/utils/teams.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+    if validate_token(response_json):
+        teams = response_json["results"]
+        if len(teams) != 0:
+            pretty_print_teams(teams)
+        else:
+            echo("Sorry, no teams found.")
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Please change it to `Sorry, no teams found!`

File: evalai/utils/teams.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+                                headers=headers,
+                                data=data
+                                )
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        if (response.status_code in EVALAI_ERROR_CODES):
+            if "team_name" in response.json().keys():
+                echo(style("Error: {}".format(response.json()["team_name"][0]), fg="red", bold=True))
+            else:
+                echo(style("Error: " + response.json()['error'], fg="red", bold=True))
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: use `.fomat` here.

File: evalai/utils/teams.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            echo(err)
+        sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        echo(err)
+        sys.exit(1)
+
+    response_json = response.json()
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: use `response_data` here.

File: evalai/utils/teams.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+    if response.status_code == 201:
+        echo(style("\nThe team {} was successfully created.\n".format(response_json["team_name"]),
+                   fg="green", bold=True))
+
+
+def challenge_participate(challenge_id, participant_team_id):
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Please change it to `participate_in_a_challenge`

File: evalai/utils/teams.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    headers['Content-Type'] = 'application/json'
+    try:
+        response = requests.post(
+                                url,
+                                headers=headers,
+                                )
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        if (response.status_code in EVALAI_ERROR_CODES):
+            echo(style("Error: " + response.json()["error"], fg="red", bold=True))
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: use `.format` here.

File: evalai/utils/teams.py [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            echo(err)
+        sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        echo(err)
+        sys.exit(1)
+
+    if response.status_code == 201:
+        echo(style("Your participant team is now participating in this challenge.",
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: Please change it to `Your team {} is now participating in this challenge!`.format(<team_name_here>)`

File: evalai/utils/urls.py [hunk 20]
```diff
@@ -9,3 +9,5 @@ class URLS(Enum):
     host_teams = "/api/hosts/challenge_host_team/"
     host_challenges = "/api/challenges/challenge_host_team/{}/challenge"
     participant_challenges = "/api/participants/participant_team/{}/challenge"
+    participant_team_lists = "/api/participants/participant_team"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py]: `participant_team_list`

File: evalai/utils/urls.py [hunk 21]
```diff
@@ -9,3 +9,5 @@ class URLS(Enum):
...
+    challenge_participate = "/api/challenges/challenge/{}/participant_team/{}"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py]: `participate_in_a_challenge`

File: tests/data/teams_response.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+
+object_error = """
+{
+    "error": "Sorry, the object does not exist."
+}
+"""
+
+
+team_exists_error = """
```
Comment (RishabhJain2018) [inline tests/data/teams_response.py]: `participant_team_already_exists_error`

File: tests/test_requests.py [hunk 23]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+    def test_participate_in_a_challenge_for_http_error_404(self):
+        runner = CliRunner()
+        result = runner.invoke(challenge, ['2', 'participate', '3'])
+        response = result.output
+        url = "{}{}".format(API_HOST_URL, URLS.challenge_participate.value).format("2", "3")
+        expected = "{}{}".format(self.expected.format(url), "\n")
+        assert response == expected
+
+
+class TestTeamsObjectDoesNotExist(BaseTestClass):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `TestTeamsWhenObjectDoesNotExist`

File: tests/test_requests.py [hunk 24]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+
+    @responses.activate
+    def test_participate_in_a_challenge_for_object_does_not_exist(self):
+        runner = CliRunner()
+        result = runner.invoke(challenge, ['2', 'participate', '3'])
+        response = result.output.rstrip()
+        assert response == self.expected
+
+
+class TestTeamsTeamNameExists(BaseTestClass):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `TestTeamsWhenTeamNameAlreadyExists`

File: evalai/main.py [hunk 25]
```diff
@@ -19,7 +19,8 @@ def main(ctx):
         echo(welcome_text)
 
 
-# Command -> evalai auth/challenges/submissions/teams
 main.add_command(challenges)
+main.add_command(challenge)
+
```
Comment (RishabhJain2018) [inline evalai/main.py]: Please remove this line.

File: evalai/challenges.py (hunk 3 above)
Comment (isht3) [reply evalai/challenges.py]: Actually, I used the current style because it did the exact functionality that you mentioned. Should I change it?

Comment (RishabhJain2018) [reply evalai/challenges.py]: Please check once. 
For more reference please see the screenshot below
screen shot 2018-06-24 at 1 07 28 am

File: tests/test_teams.py [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+            team_name = "\n{}".format(team["team_name"],)
+            team_id = "ID: {}\n\n".format(str(team["id"]),)
+            title = "{} {}".format(team_name, team_id)
+            created_by = "Created by : {}\n\n".format(team["created_by"])
+            members = "{}\n".format("Members")
+            for member in team["members"]:
+                members = "{}* {}\n".format(members, member["member_name"])
+            team = "{}{}{}\n{}\n".format(title, created_by, members, br)
+
+            output = output + team
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: use `.format` here.

File: tests/test_requests.py [hunk 27]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
         url = "{}{}".format(API_HOST_URL, URLS.host_teams.value)
         assert response == self.expected.format(url)
 
+    @responses.activate
+    def test_display_teams_lists_for_http_error_404(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_display_participant_team_for_http_error_404`

File: tests/test_requests.py [hunk 28]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+        runner = CliRunner()
+        result = runner.invoke(teams)
+        response = result.output
+        url = "{}{}".format(API_HOST_URL, URLS.participant_team_list.value)
+        expected = "{}{}".format(self.expected.format(url), "\n")
+        assert response == expected
+
+    @responses.activate
+    def test_create_team_for_http_error_404(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_create_participant_team_for_http_error_404`

File: evalai/utils/teams.py [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+        team_name = "{} {}".format(team_name, team_id)
+        created_by = "Created by : {}\n\n".format(style(team["created_by"], fg="yellow"))
+        members = "{}\n".format(style("Members", bold=True))
+        for member in team["members"]:
+            members = "{}* {}\n".format(members, member["member_name"])
+        team = "{}{}{}\n{}".format(team_name, created_by, members, br)
+        echo(team)
+
+
+def display_participant_teams():
```
Comment (RishabhJain2018) [inline evalai/utils/teams.py]: `display_participant_teams`

File: tests/test_requests.py [hunk 30]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+
+        responses.add(responses.POST, url.format(API_HOST_URL, URLS.participant_team_list.value), json=error_data,
+                      status=406)
+
+        responses.add(responses.POST, url.format(API_HOST_URL, URLS.participate_in_a_challenge.value).format("2", "3"),
+                      json=error_data, status=406)
+        self.expected = "Error: Sorry, the object does not exist."
+
+    @responses.activate
+    def test_display_teams_lists_for_object_does_not_exist(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_display_participant_team_for_object_does_not_exist`

File: tests/test_requests.py [hunk 31]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+        runner = CliRunner()
+        result = runner.invoke(teams)
+        response = result.output.rstrip()
+        assert response == self.expected
+
+    @responses.activate
+    def test_create_team_for_object_does_not_exist(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_create_participant_team_for_object_does_not_exist`

File: tests/test_requests.py [hunk 32]
```diff
@@ -91,6 +104,102 @@ def test_display_participant_and_host_challenge_lists_for_http_error_404(self):
...
+        runner = CliRunner()
+        result = runner.invoke(teams, ['create'], input="TeamTest\ny\n")
+        response = result.output.rstrip()
+        user_prompt_text = ("Enter team name: : TeamTest\n"
+                            "Please confirm the team name - TeamTest [y/N]: y\n")
+        expected = "{}{}".format(user_prompt_text, self.expected)
+        assert response == expected
+
+    @responses.activate
+    def test_participate_in_a_challenge_for_object_does_not_exist(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_participate_in_a_challenge_when_object_does_not_exist`

File: tests/test_requests.py [hunk 33]
```diff
@@ -202,3 +324,21 @@ def test_display_participant_and_host_challenge_lists_for_request_exception(self
         runner = CliRunner()
         result = runner.invoke(challenges, ['--participant', '--host'])
         assert result.exit_code == -1
+
+    @responses.activate
+    def test_display_teams_lists_for_request_exception(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_display_participant_team_for_request_exception`

File: tests/test_requests.py [hunk 34]
```diff
@@ -202,3 +324,21 @@ def test_display_participant_and_host_challenge_lists_for_request_exception(self
...
+        runner = CliRunner()
+        result = runner.invoke(teams)
+        assert result.exit_code == -1
+
+    @responses.activate
+    def test_create_team_for_request_exception(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: `test_create_participant_team_for_request_exception`

File: tests/test_teams.py [hunk 35]
```diff
@@ -0,0 +1,117 @@
...
+    def test_teams_list(self):
+
+        output = ""
+        for team in self.teams:
+            br = ("----------------------------------------"
+                  "--------------------------")
+
+            team_name = "\n{}".format(team["team_name"],)
+            team_id = "ID: {}\n\n".format(str(team["id"]),)
+            title = "{} {}".format(team_name, team_id)
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: Please change `title` -> `team_name`

File: tests/test_teams.py [hunk 36]
```diff
@@ -0,0 +1,117 @@
...
+
+            output = "{}{}".format(output, team)
+
+        runner = CliRunner()
+        result = runner.invoke(teams)
+        response = result.output
+        assert response == output
+
+    @responses.activate
+    def test_create_team_for_option_yes(self):
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: `test_create_participant_team_for_option_yes`

File: tests/test_teams.py [hunk 37]
```diff
@@ -0,0 +1,117 @@
...
+        output = ("Enter team name: : TeamTest\n"
+                  "Please confirm the team name - TeamTest [y/N]: y\n"
+                  "\nThe team TestTeam was successfully created.\n\n")
+        runner = CliRunner()
+        result = runner.invoke(teams, ['create'], input="TeamTest\ny\n")
+        response = result.output
+        assert response == output
+
+    @responses.activate
+    def test_create_team_for_option_no(self):
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: `test_create_participant_team_for_option_no`

File: tests/test_teams.py [hunk 38]
```diff
@@ -0,0 +1,117 @@
...
+        output = ("Enter team name: : TeamTest\n"
+                  "Please confirm the team name - TeamTest [y/N]: n\n"
+                  "Aborted!\n")
+        runner = CliRunner()
+        result = runner.invoke(teams, ['create'], input="TeamTest\nn\n")
+        response = result.output
+        assert response == output
+
+    @responses.activate
+    def test_participate(self):
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: `test_participate_in_a_challenge`

File: tests/test_teams.py [hunk 39]
```diff
@@ -0,0 +1,117 @@
...
+        output = "Your team id {} is now participating in this challenge!\n".format("3")
+        runner = CliRunner()
+        result = runner.invoke(challenge, ['2', 'participate', '3'])
+        response = result.output
+        assert response == output
+
+    @responses.activate
+    def test_participate_with_single_argument(self):
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: `test_participate_in_a_challenge_with_single_argument`

File: tests/test_teams.py [hunk 40]
```diff
@@ -0,0 +1,117 @@
...
+        output = ("Usage: challenge [OPTIONS] CHALLENGE COMMAND [ARGS]...\n"
+                  "\nError: Missing command.\n")
+        runner = CliRunner()
+        result = runner.invoke(challenge, ['2'])
+        response = result.output
+        assert response == output
+
+    @responses.activate
+    def test_participate_with_string_argument(self):
```
Comment (RishabhJain2018) [inline tests/test_teams.py]: `test_participate_in_a_challenge_with_string_argument`

File: evalai/utils/teams.py (hunk 29 above)
Comment (isht3) [reply evalai/utils/teams.py]: It's the same, isn't it?

Comment (RishabhJain2018) [reply evalai/utils/teams.py]: yes..yes! I looks fine!.

**Pull request merged**